### PR TITLE
SI-6807 Deprecating the Actors library.

### PR DESCRIPTION
**Pull request by vjovanov**
All public classes, traits and objects marked as deprecated.
Added deprecation note on the package object.
Embedded external libraries (ThreadPool etc.) are not deprecated as they are intended for internal use only.

Review by: @phaller

### CODE REVIEW

**Comment by retronym** [comment]
I don't remember the discussions on this, but what's the rationale behind deprecating in 2.10.1? Is the plan to remove this in 2.12.0? We might also be able to farm it out to a separate repository in the meantime, so long as we include it with the distribution.

**Comment by soc** [comment]
LGTM. Somehow I get the impression that we forgot to deprecate tons of stuff for 2.10.0 ...

**Comment by vjovanov** [comment]
We deprecate them in 2.10.1 because users should stop using them as soon as possible. Unfortunately I missed the opportunity to deprecate Actors during the RC phase. Roland noticed that actors are not deprecated during RC5 when it was already too late. Based on the user feedback on the migration kit we should decide when to remove Actors from the distribution.

**Comment by adriaanm** [comment]
@soc, do you have a list in mind?

@vjovanov, I agree this is the right thing to do -- better late than never :-)

**Comment by gkossakowski** [comment]
@adriaanm: that's not true. We should not deprecate things in minor releases. Minor releases are supposed to be drop-in replacements offerring painless upgrade. There are many projects that are very strict about using deprecated stuff to the point that they make their builds fail if you happen to use deprecated classes or methods. If we deprecate things in minor releases we might be breaking people's builds.

**Comment by adriaanm** [comment]
well -- we may have made a mistake not deprecating as much as we should have
what's better? 
- sticking with that mistake and not being able to remove stuff in 2.11?
- or admitting the mistake, causing some pain in upgrades to 2.10.1, but being able to slim down 2.11 more?

**Comment by adriaanm** [comment]
hmmm... I'm not convinced that many people compile with -Xfatal-warnings as we generate so many of them,
so I don't think there are many scala projects out there whose build will be broken by new deprecations

of course it would have been better to have them in 2.10.0, but how many people will be waiting for 2.10.1 anyway?

**Comment by gkossakowski** [comment]
@adriaanm: I saw at least one project (not publicly available) where it was scrapping compiler's logging output to detect deprecation warnings only and make them fatal.

Options we have:
- not deprecate in 2.10.1 and deprecate in 2.11.0
- deprecate in 2.10.1 what we absolutely need (like actors) and make it _very_ prominent in release notes (we should have a ticket for that so it's not forgotten) that we did this and that was an exception and we are not planning to deprecate anything further in 2.10.x series

It's not I'm absolutely against fixing a mistake but that decisions shouldn't be taken lightly and we should be aware of tradeoffs (like breaking people's builds, for real).

**Comment by adriaanm** [comment]
you're right -- this is not a decision to be taken lightly, but I think it's worthwhile to come up with a list of stuff we really want to remove in 2.11 but forgot to deprecate in 2.10.0
- actors
- swing?
- ...

to be continued on scala-internals...

**Comment by paulp** [comment]
```
Yes, swing, actors, and: scala.xml.*, specialization, continuations plugin, lots more
@deprecatedInheritance, scala.math.*, Synchronized*, *Proxy*, *View*, *Forwarder*,
scala.collection.parallel.*, scala.collection.*Gen*, for starters. I'd deprecate as much
as I could, and push the rest outside the standard library.
```

**Comment by soc** [comment]
@adriaanm I think that's the right thread: https://groups.google.com/d/topic/scala-internals/axzOQ6fUNQg/discussion

File: src/actors/scala/actors/Future.scala [hunk 1]
```diff
@@ -21,6 +21,7 @@ import scala.concurrent.SyncVar
  *
  *  @author Philipp Haller
  */
+@deprecated("Use the akka.actor package instead. For migration from the scala.actors package refer to the Actors Migration Guide.", "2.10.1")
```
Comment (phaller) [inline src/actors/scala/actors/Future.scala:24]: Exceptionally, I think here we should refer to `scala.concurrent.Future`:
"Use scala.concurrent.Future instead"

File: src/actors/scala/actors/Future.scala [hunk 2]
```diff
@@ -107,6 +108,7 @@ private class FutureActor[T](fun: SyncVar[T] => Unit, channel: Channel[T]) exten
...
```
Comment (phaller) [inline src/actors/scala/actors/Future.scala:111]: Again, here we should probably say "Use object scala.concurrent.Future instead"

**Comment by phaller** [comment]
Apart from my two nitpicks in the commit, it looks good. Thanks!

**Comment by adriaanm** [comment]
This deprecation should happen in 2.11.0, as discussed in the scala meeting.

**Pull request closed without merging**